cpp/src/arrow/util/async_generator_test.cc

```diff
@@ -1095,40 +1095,51 @@ TEST(TestAsyncUtil, ReadaheadMove) {
 TEST(TestAsyncUtil, ReadaheadFailed) {
   ASSERT_OK_AND_ASSIGN(auto thread_pool, internal::ThreadPool::Make(4));
   std::atomic<int32_t> counter(0);
+  std::mutex mx;
+  bool error_unlocked = false;
```

Comment:
Is it possible to make this test simpler?
I think this test is to verify failed readahead is detected, and won't interfere with successful readaheads.
Can we keep original code, with minor changes to make sure `counter` is incrementing till `IterationEnd`, but allow `counter` to be large?

Reply:
Thanks for the input.  I made a pass at simplifying.  I want to make sure there are at least some failed tasks so I left in a bit of gating (but reused a helper GatingTask instead of using mutexes and condition variables).  I did change the counter at the end to accept any # of remaining tasks.